### DISCUSSION TITLE
OCILIB 6.0: Survey and call to ideas for library updates

### DISCUSSION THREAD

**Discussion by vrogier** [vrogier/ocilib#269] [Ideas]
Dear all,

OCILIB is almost 10 years old :) 

Since version 1.0.0, I've always preserved backward compatibility across new revisions of the library.
It was a lot of constraints to achieve it and sometimes new additions were not designed right when trying to fit into the existing API.

But at some point, the API needs to be changed to get better.
I've decided to introduce breaking changes for version 6.0 ([edit 2018-10-30]~~before the end of the year~~ mid 
 2019).

[edit 2018-10-30] version 5.0 will be a C++ API refactoring.

Current version will move to a v4 branch and will still be maintained and updated!

I would like gather your opinion about what needs to be changed, redesigned or added.

Version 6 will not be a full rewrite nor a massive change in the API but rather fixing incoherence,  making better what is not done right and adding what's missing.

For example:

- for the the following types, constructor methods take an OCI_Connection argument that is useless and the argument should be removed: OCI_Number, OCI_Date, OCI_Timestamp, OCI_Interval, OCI_Object, OCI_Ref
- Function naming is sometimes lacking of coherence
- OCI_BindXXX() calls should return an OCI_Bind object instead of returning boolean
- OCI_BindXXX() calls should take as parameter the bind direction (in, out, in/out) instead of forcing applications to call OCI_GetBind() in order to call OCI_BindSetDirection()
- Providing sizes for arrays for DML  and PL/SQL tables uses different mechanism 
- ...

I am building my own todo list from my experience over the years.

Please, any insights, ideas and proposals are welcome :)

Thank you

Vincent

**Comment by bangusi** [comment]
It is not broken. No need to fix.
Seriously I am new to ocilib. Only became aware of it last year I think but started using it a couple of months ago. I am greatly impressed.

**Comment by vrogier** [comment]
Hi all,

Is there anybody having concerns regarding me porting the C implementation to C++ ?
OCILIB would still export its C API as before and will have its C++ API on top of the C API.
No public changes.
For quite a while I really wanted to switch to C++ for the core implementation.
Main reasons are:
- no more macros (ugly, factorized but duplicated code, not debuggable, ...)
- drastic codebase size decrease (using STL and templated code) at least by factor x2
- possible performance gain (C++ compiler optimization and inlining) 
- more readable code 
- more compact code
- ....

There is no way that I would start writing OCILIB from scratch in plain C these days. 
OCILIB is the only piece of code I still code in C.
I do not see any good reasons to keep its implementation in C if I still provide the C interface.

Regards,

Vincent

**Comment by bangusi** [comment]
It is your library so obviously you can do whatever you please.
That said, if you go this route I may have to re-write or at least tweak my program in which I am actually using the C API in a  C++ program. It is my very first OCILIB program and  I really did not find a compelling reason to use the C++ API which is a wrapper over the C API which is itself a wrapper over OCI.  My thinking was that the C++ must have some overhead compared to the C API so why bother when the C API is perfectly usable from C++.

**Comment by vrogier** [comment]
@bangusi 

Implementing public C API using  C++ internally does not change anything for you.
If I hadn't yet moved internal implementation to C++, it is because I wanted to have native C++ implementation used for both C++ public API and public C API implementation.

But It takes quite a lot of time to move on as it is a major change /rewrite. And I am fed up to delay that move.

Instead for version 6.0, I am thinking to first implement the public C API using C++ implementation (to make the code base much smaller, easier to maintain, ad possibly more efficient with less memory allocations) and keep the current C++ API on top of the C API.

Once this move done, it would be much easier to upgrade/replace the C++ internal implementation with a one using a C++ public API.
In your case, It would eliminate a (really thin) layer from the stack.

This said, current overhead of the C++ API is very very small.  It is more  like sugar syntax on the C API.
From my testing and benchmarks, there is no noticeable overhead over given workflows / use cases.
The only thing that could come to my mind is related to std::string allocations.

I've coded many similar OCI and OCILIB C/C++ programs .
OCILIB C API is a (really nice) sugar syntax on top of OCI.
It has no overhead as most of its code is about performing tasks you would have to code yourself in OCI programs. Of course the best C programmer in the world may write more efficient code than I did, but at what cost and maintainability price ? The only OCILIB overhead is about argument safety (C API validates all arguments) which I think is reasonable tiny price to pay (one thing I could make optional at build time when porting to C++ internal implementation)
OCILIB C++ API is the same stuff on top of it C API. 
Tiny overhead for std::string allocations and vector binding management.
More than 99.9% execution time of applications using OCILIB is either in user code or in OCI code. 
Not OCILIB code.

**Comment by andygeng** [comment]
Hi Vincent,
I didn't see xmltype demos, does ocilib support it?
Thanks.

**Comment by dbpm** [comment]
vrogier

native C++ implementation is unwanted because it will lead to excessive linkage of C++ system libraries. It's definitely not LINUX/UNIX way (where C++ is not being used by default).

For an example, completely empty program likee

    user@server.intranet:~#cat test.cpp
    int main() {
            return 0;
    }

compiled by C++ compiler immediately gives additional dependencies (libstdc++, libm, libgcc_s):

    user@server.intranet:~#g++ test.cpp
    user@server.intranet:~#ldd a.out
          linux-vdso.so.1 =>  (0x00007fff57db2000)
          libstdc++.so.6 => /lib64/libstdc++.so.6 (0x00007f6a239c9000)
          libm.so.6 => /lib64/libm.so.6 (0x00007f6a236c7000)
          libgcc_s.so.1 => /lib64/libgcc_s.so.1 (0x00007f6a234b0000)
          libc.so.6 => /lib64/libc.so.6 (0x00007f6a230ed000)
          /lib64/ld-linux-x86-64.so.2 (0x0000560ecf4d4000)

while C one - does not:

    user@server.intranet:~#gcc test.cpp
    user@server.intranet:~#ldd a.out
        linux-vdso.so.1 =>  (0x00007ffe901f5000)
        libc.so.6 => /lib64/libc.so.6 (0x00007f0932b08000)
        /lib64/ld-linux-x86-64.so.2 (0x0000559bb3932000)

Oracle's libraries like Instant client also do not have any dependency on C++ stuffs.

If new OCILIB will be completely rewritten on C++ I personally will have to branch out 4.x and do not use 6.x, because fighting with bugs in non GCC/Clangs compilers like old Sunpro ones is not an option for me, as well as wasting time to get expertise in modern C++.

**Comment by bangusi** [comment]
The C++ compiler issue is definitely an important one. Once the API is C++ only we have to care which compiler was used to create ocilib.dll. I don't use Oracle's OCCI for this reason. You are forced choose a compiler that they used and this is not always ideal. Of course since the source for OCILIB is open you can let users compile their own

**Comment by dbpm** [comment]
As alternative and as probably most proper approach it will be better to split OCILIB into OCILIB, and let's say OCILIB++
Two completely different products/libraries - one will be Pure ANSI C implementation and second one will be STL based C++ implementation, each one with own lifecycle (releases/bug fixes/etc).

From Pure C developer point of view – all those shipped .cpp files are useless - I do nothing than removing them from library linking.
Meanwhile current C implementation looks very good, your C coding style is much more than impressive, the macros part especially and it will be too sad to loss it.

From other hand, from typical C++ developer’s point of view, all those OCILIB C specific lists, exceptions, string handling routines look like redundant reinvention of the wheel, indeed, so pure STL based C++ reimplementation is much more than reasonable.

But I doubt that compromise and single codebase is even possible there - Pure C and C++ worlds and principles are too far from each other

**Comment by vrogier** [comment]
@bangusi @dbpm 
About dependencies, it depends how your compile the library. 
It could not have any dependencies at all if needed. 
For example, On Windows, OCILIB DLLs statically link the MSVC runtime and thus do not depend on any version of it nor any redistribuables. The only dependency is kernel32.dll
Regarding C++, same, it also can be linked statically without any dependencies.

@dbpm 
Here are ldd outputs from both OCILIB C and C++ demo applications:

```bash
[vincent@localhost test]$ ldd ocilib_demo++
	libgcc_s.so.1 => /lib64/libgcc_s.so.1 (0x000000324bc00000)
	libstdc++.so.6 => /usr/lib64/libstdc++.so.6 (0x000000324c000000)
	linux-vdso.so.1 =>  (0x00007fff781ff000)
	libclntsh.so.11.1 => /usr/lib/oracle/11.2/client64/lib/libclntsh.so.11.1 (0x00007f62b8c06000)
	libnnz11.so => /usr/lib/oracle/11.2/client64/lib/libnnz11.so (0x00007f62b85b0000)
	libocilib.so.4 => /usr/local/lib/libocilib.so.4 (0x00007f62b898b000)
	libm.so.6 => /lib64/libm.so.6 (0x00000037af400000)
	libc.so.6 => /lib64/libc.so.6 (0x00000037ae800000)
	libpthread.so.0 => /lib64/libpthread.so.0 (0x00000037af000000)
	libdl.so.2 => /lib64/libdl.so.2 (0x00000037aec00000)
	libnsl.so.1 => /lib64/libnsl.so.1 (0x00000037bd000000)
	libaio.so.1 => /lib64/libaio.so.1 (0x0000003e1d800000)
	/lib64/ld-linux-x86-64.so.2 (0x00000037ae400000)

[vincent@localhost test]$ ldd ocilib_demo
	linux-vdso.so.1 =>  (0x00007ffffc1ff000)
	libclntsh.so.11.1 => /usr/lib/oracle/11.2/client64/lib/libclntsh.so.11.1 (0x00007f838e107000)
	libnnz11.so => /usr/lib/oracle/11.2/client64/lib/libnnz11.so (0x00007f838dab3000)
	libocilib.so.4 => /usr/local/lib/libocilib.so.4 (0x00007f838de8c000)
	libm.so.6 => /lib64/libm.so.6 (0x00000037af400000)
	libc.so.6 => /lib64/libc.so.6 (0x00000037ae800000)
	libpthread.so.0 => /lib64/libpthread.so.0 (0x00000037af000000)
	libdl.so.2 => /lib64/libdl.so.2 (0x00000037aec00000)
	libnsl.so.1 => /lib64/libnsl.so.1 (0x00000037bd000000)
	libaio.so.1 => /lib64/libaio.so.1 (0x0000003e1d800000)
	/lib64/ld-linux-x86-64.so.2 (0x00000037ae400000)

```

only added dependencies are libgcc_s.so.1 and libstdc++.so.6 (2 out of 11 ones).

I do not see drawbacks having a shared library exposing a C interface that has internal implementation in C++. Except not being able to dig into the code when not knowing C++ (would the same for any language producing binary code).

That's funny how C++ is seen as "the evil" by C developers. I will not get into that kind of discussion here but so far, all of "anti C++" arguments I have heard the past few years are based on biased clichés.

But If the OCILIB community prefer that OCILIB source code remains in C, I will not go against it.
I will then keep current C implementation (and feel pain writing/maintaining it).
I guess that the price to pay for people to use the library.
While it could be drastically enhanced, shortened, made safer and more performant with C++.

Best regards,

Vincent

**Comment by vrogier** [comment]
Hi,

I've started to work on version 6.0.
I will keep C as implementation language and will proceed to important internal refactoring and API breaking changes.

Regards,

Vincent

**Comment by bangusi** [comment]
@vrogier Looking forward to a new and improved ocilib. Let me know whenever you want someone to kick the tires

**Comment by vrogier** [comment]
@bangusi  let me know what stuff you want to be added and or improved :)

**Comment by bangusi** [comment]
@vrogier  Since you asked.. I would like to see an easier ( more intuitive ) way to do array binding.
A user provides array size, and commit interval. The library takes care of building the array. resizing, commit etc. I looked at the internal interface but it didn't seem to provide much better abstraction than the regular interface.

**Comment by vrogier** [comment]
Just for information.
v5.0 will be pushed backed to 2019.
In the meanwhile, I will focus on v4.6 that will add support for Oracle 18 new features.
